src/algebra/module/pid.lean

```diff
@@ -0,0 +1,300 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import algebra.module.torsion
+import linear_algebra.free_module.pid
+import algebra.module.projective
+
+/-!
+# Structure of finitely generated modules over a PID
+
+## Main statements
+
+* `equiv_direct_sum_of_is_torsion` : A finitely generated torsion module over a PID is isomorphic
+  to a direct sum of some `R ⧸ R ∙ (p i ^ e i)` where the `p i ^ e i` are prime powers.
+* `equiv_free_prod_direct_sum` : A finitely generated module over a PID is isomorphic to the
+  product of a free module (its torsion free part) and a direct sum of the form above (its torsion
+  submodule).
+
+## Notation
+
+* `R` is a PID and `M` is a (finitely generated for main statements) torsion module over R.
+* `p` is an irreducible element of `R` or a tuple of these.
+
+## Implementation details
+
+We first prove (`is_internal_prime_power_torsion`) a finitely generated torsion module is the
+internal direct sum of its `p i ^ e i`-torsion submodules for some (finitely many) prime powers
+`p i ^ e i`.
+
+Then we treat the case of a `p ^ ∞`-torsion module (that is, a module where all elements are
+cancelled by scalar multiplication by some power of `p`) and apply it to the `p i ^ e i`-torsion
+submodules (that are `p i ^ ∞`-torsion) to get the result for torsion modules.
+
+Then we get the general result using that a torsion free module is free (which has been proved at
+`module.free_of_finite_type_torsion_free'` at `linear_algebra/free_module/pid.lean`.)
+
+## Tags
+
+Finitely generated module, principal ideal domain, classification, structure theorem
+-/
+
+universes u v
+open_locale big_operators
+
+section split_exact --where to move ??
```

Comment:
I think we should use the general results in #14009 , once it gets merged.

Reply:
It seems so far we only have the results for left split sequences, and in my case it's actually right split

Reply:
Hmm, this is annoying, but I really think we should try to develop the theory in general. Maye you could try to dualize the results in #14009. If you're not familiar with the category theory part of the library I am sure you'll find help on zulip.

Reply:
So I managed to prove it using the results about split exact sequences (I wrote and pushed the "right split exact" versions of the lemmas), but only when A, M and B are on the same universe (I didn't push it because it breaks the following proofs)

Reply:
As I was able to adapt the proof of the final theorem using ulift, I pushed a version with just the specific lemma on split exact sequences being universe-restricted

Reply:
Would you mind opening another PR with the material about split exact sequence? Just to keep this one smaller. I promise a quick review!

Reply:
Done at #14376